src/libraries/System.Private.CoreLib/src/System/IO/AsyncWindowsFileStreamStrategy.cs

```diff
@@ -0,0 +1,406 @@
+        {
+            // This is necessary for async IO using IO Completion ports via our
+            // managed Threadpool API's.  This (theoretically) calls the OS's
+            // BindIoCompletionCallback method, and passes in a stub for the
+            // LPOVERLAPPED_COMPLETION_ROUTINE.  This stub looks at the Overlapped
+            // struct for this request and gets a delegate to a managed callback
+            // from there, which it then calls on a threadpool thread.  (We allocate
+            // our native OVERLAPPED structs 2 pointers too large and store EE state
+            // & GC handles there, one to an IAsyncResult, the other to a delegate.)
+            try
+            {
+                _fileHandle.ThreadPoolBinding = ThreadPoolBoundHandle.BindHandle(_fileHandle);
+            }
+            catch (ArgumentException ex)
+            {
+                throw new IOException(SR.IO_BindHandleFailed, ex);
+            }
+            finally
+            {
+                if (_fileHandle.ThreadPoolBinding == null)
+                {
+                    // We should close the handle so that the handle is not open until SafeFileHandle GC
+                    Debug.Assert(!_exposedHandle, "Are we closing handle that we exposed/not own, how?");
+                    _fileHandle.Dispose();
+                }
+            }
+        }
+
+        // called by BufferedStream
+        internal void OnBufferAllocated(byte[] buffer)
+        {
+            Debug.Assert(buffer != null);
+            Debug.Assert(_preallocatedOverlapped == null);
+
+            _preallocatedOverlapped = new PreAllocatedOverlapped(FileStreamCompletionSource.s_ioCallback, this, buffer);
+        }
+
+        SafeFileHandle IFileStreamCompletionSourceStrategy.FileHandle => _fileHandle;
+
+        FileStreamCompletionSource? IFileStreamCompletionSourceStrategy.CurrentOverlappedOwner => _currentOverlappedOwner;
+
+        FileStreamCompletionSource? IFileStreamCompletionSourceStrategy.CompareExchangeCurrentOverlappedOwner(FileStreamCompletionSource? newSource, FileStreamCompletionSource? existingSource)
+            => Interlocked.CompareExchange(ref _currentOverlappedOwner, newSource, existingSource);
+
+        public override int Read(byte[] buffer, int offset, int count)
+            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count)).GetAwaiter().GetResult();
+
+        public override Task<int> ReadAsync(byte[] buffer, int offset, int count, CancellationToken cancellationToken)
+            => ReadAsyncInternal(new Memory<byte>(buffer, offset, count), cancellationToken);
+
+        public override ValueTask<int> ReadAsync(Memory<byte> destination, CancellationToken cancellationToken = default)
+            => new ValueTask<int>(ReadAsyncInternal(destination, cancellationToken));
+
+        private unsafe Task<int> ReadAsyncInternal(Memory<byte> destination, CancellationToken cancellationToken = default)
```

Comment:
Great for now as part of the refactoring.  Subsequently we'll want to look into making FileStreamCompletionSource be a reusable IValueTaskSource implementation, and have this method return ValueTask instead of Task.  At least with how I'm thinking of use cases, that's going to be most impactful when not buffering.

Reply:
cc @carlossanlop who will work on reducing the allocations